### PR TITLE
chore(auth): Unit tests for fetchCurrentDevice API

### PR DESCRIPTION
**Pull request by hahnandrew**
*Description of changes:*

Adding unit tests for `fetchCurrentDevice` 

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

### CODE REVIEW

File: packages/auth/amplify_auth_cognito_test/test/plugin/fetch_current_device.dart [hunk 1]
```diff
@@ -0,0 +1,122 @@
+// Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
```
Comment (Jordan-Nelson) [inline packages/auth/amplify_auth_cognito_test/test/plugin/fetch_current_device.dart:1]: *todo:* Rename the file to end in `_test.dart`

`dart test` by default looks for file names that match the pattern `*_test.dart` (see: https://pub.dev/packages/test#running-tests)

File: packages/auth/amplify_auth_cognito_test/test/plugin/fetch_current_device.dart [hunk 2]
```diff
@@ -0,0 +1,122 @@
...
+  final userPoolKeys = CognitoUserPoolKeys(userPoolConfig);
+  final identityPoolKeys = CognitoIdentityPoolKeys(identityPoolConfig);
+  final testAuthRepo = AmplifyAuthProviderRepository();
+  final mockDevice = DeviceType(deviceKey: deviceKey);
+  final mockDeviceResponse = GetDeviceResponse(device: mockDevice);
+
+  late DeviceMetadataRepository repo;
+  late AmplifyAuthCognitoDart plugin;
+  late CognitoAuthStateMachine stateMachine;
+  late MockSecureStorage secureStorage;
```
Comment (Jordan-Nelson) [inline packages/auth/amplify_auth_cognito_test/test/plugin/fetch_current_device.dart:27]: *suggestion:* Move this inside the two `setUp` fns where it is initialized and make it final.

Since you are not referencing this outside of `setUp` functions, this can be local to those functions.

Comment (Jordan-Nelson) [reply packages/auth/amplify_auth_cognito_test/test/plugin/fetch_current_device.dart:27]: The same applies for stateMachine

File: packages/auth/amplify_auth_cognito_test/test/plugin/fetch_current_device.dart [hunk 3]
```diff
@@ -0,0 +1,122 @@
...
+        deviceKeys: CognitoDeviceKeys(userPoolConfig, username),
+      );
+      plugin = AmplifyAuthCognitoDart(
+        secureStorageFactory: (_) => secureStorage,
+      );
+      stateMachine = plugin.stateMachine;
+      await plugin.configure(
+        config: mockConfig,
+        authProviderRepo: testAuthRepo,
+      );
```
Comment (Jordan-Nelson) [inline packages/auth/amplify_auth_cognito_test/test/plugin/fetch_current_device.dart:101]: *nit:* You could move the common setup steps to a single group/setUp fn.

```dart
group('fetchCurrentDevice', () {
  setUp(() async { /* common setup */ });
  group('group 1', () {
    setUp(() async { /* group 1 setup */ });
    // tests
  });
  group('group 2', () {
    setUp(() async { /* group 2 setup */ });
    // tests
  });
});

File: packages/auth/amplify_auth_cognito_test/test/plugin/fetch_current_device.dart (hunk 1 above)
Comment (hahnandrew) [reply packages/auth/amplify_auth_cognito_test/test/plugin/fetch_current_device.dart]: added

File: packages/auth/amplify_auth_cognito_test/test/plugin/fetch_current_device.dart (hunk 2 above)
Comment (hahnandrew) [reply packages/auth/amplify_auth_cognito_test/test/plugin/fetch_current_device.dart]: I can't apply both this feedback and your [other one](https://github.com/aws-amplify/amplify-flutter/pull/5074#discussion_r1661420403) with Dart's scoping. 

Variables defined in a nested `setUp()`'s async body aren't scoped to its corresponding `group()` by Dart. That is, if a parent `setUp()` has async declarations, its children `setUp()`'s aren't able to access that scope without the top level late declaration. I can either have a global `setUp()` with declarations late declared at top level, or decouple the `setUp()` function per `group()` without the top level late declarations. 

Which feedback should I prioritize? @Jordan-Nelson

File: packages/auth/amplify_auth_cognito_test/test/plugin/fetch_current_device.dart (hunk 3 above)
Comment (hahnandrew) [reply packages/auth/amplify_auth_cognito_test/test/plugin/fetch_current_device.dart]: See [response](https://github.com/aws-amplify/amplify-flutter/pull/5074#discussion_r1661553305)

tldr; The top level late declaration gives async children `setUp()`'s access to the asynchronously resolved declarations in its parent `setUp()`'s body. Without it, I can't have a common `setUp()`, and need to decouple and re-declare the mocks.

File: packages/auth/amplify_auth_cognito_test/test/plugin/fetch_current_device.dart (hunk 2 above)
Comment (hahnandrew) [reply packages/auth/amplify_auth_cognito_test/test/plugin/fetch_current_device.dart]: > Since you are not referencing this outside of `setUp` functions

I do - I have to reference them in the tests I'm setting them up for. `setUp()` and `test()` have access to same `group` closure but not the same scope, just like the pseudocode example you wrote [here](https://github.com/aws-amplify/amplify-flutter/pull/5074#discussion_r1661420403).

Comment (Jordan-Nelson) [reply packages/auth/amplify_auth_cognito_test/test/plugin/fetch_current_device.dart]: > I do reference plugin in the tests I'm setting them up for.

I was referring to secureStorage which is defined on the line that I left a comment on.

Here is how you can incorporate both pieces of feedback that I left. I have refactored the tests to consolidate to common setup and tear down. I moved the declaration for secureStorage into the setup fn and made it final. I removed the declaration for stateMachine as you can refer to it by plugin.stateMachine. 

```dart
// Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
// SPDX-License-Identifier: Apache-2.0

import 'package:amplify_auth_cognito_dart/amplify_auth_cognito_dart.dart';
import 'package:amplify_auth_cognito_dart/src/credentials/cognito_keys.dart';
import 'package:amplify_auth_cognito_dart/src/credentials/device_metadata_repository.dart';
import 'package:amplify_auth_cognito_dart/src/sdk/cognito_identity_provider.dart';
import 'package:amplify_auth_cognito_test/common/mock_clients.dart';
import 'package:amplify_auth_cognito_test/common/mock_config.dart';
import 'package:amplify_auth_cognito_test/common/mock_secure_storage.dart';
import 'package:amplify_core/amplify_core.dart';
import 'package:test/test.dart';

void main() {
  AmplifyLogger().logLevel = LogLevel.verbose;

  final userPoolKeys = CognitoUserPoolKeys(userPoolConfig);
  final identityPoolKeys = CognitoIdentityPoolKeys(identityPoolConfig);
  final testAuthRepo = AmplifyAuthProviderRepository();
  final mockDevice = DeviceType(deviceKey: deviceKey);
  final mockDeviceResponse = GetDeviceResponse(device: mockDevice);

  late DeviceMetadataRepository repo;
  late AmplifyAuthCognitoDart plugin;

  Future<String?> getDeviceKey() async {
    final secrets = await repo.get(username);
    return secrets?.deviceKey;
  }

  group('fetchCurrentDevice', () {
    setUp(() async {
      final secureStorage = MockSecureStorage();
      seedStorage(
        secureStorage,
        userPoolKeys: userPoolKeys,
        identityPoolKeys: identityPoolKeys,
        deviceKeys: CognitoDeviceKeys(userPoolConfig, username),
      );
      plugin = AmplifyAuthCognitoDart(
        secureStorageFactory: (_) => secureStorage,
      );
      await plugin.configure(
        config: mockConfig,
        authProviderRepo: testAuthRepo,
      );
      repo = plugin.stateMachine.getOrCreate<DeviceMetadataRepository>();
    });

    group('Cognito GetDevice returns successfully', () {
      setUp(() async {
        final mockIdp = MockCognitoIdentityProviderClient(
          getDevice: () async => mockDeviceResponse,
          forgetDevice: () async {},
        );
        plugin.stateMachine.addInstance<CognitoIdentityProviderClient>(mockIdp);
      });

      test(
          'should get the current device. current device id should be equal to the local device id',
          () async {
        final currentDeviceKey = await getDeviceKey();
        expect(currentDeviceKey, isNotNull);
        final currentDevice = await plugin.fetchCurrentDevice();
        expect(currentDeviceKey, currentDevice.id);
      });

      test(
          'should should throw a DeviceNotTrackedException when current device key is null',
          () async {
        await plugin.forgetDevice();
        await expectLater(
          plugin.fetchCurrentDevice,
          throwsA(isA<DeviceNotTrackedException>()),
        );
      });
    });

    group('Cognito GetDevice throws AWSHttpException', () {
      setUp(() async {
        final mockIdp = MockCognitoIdentityProviderClient(
          getDevice: () async => throw AWSHttpException(
            AWSHttpRequest.get(Uri.parse('https://www.amazon.com')),
          ),
        );
        plugin.stateMachine.addInstance<CognitoIdentityProviderClient>(mockIdp);
      });

      test('should throw a NetworkException', () async {
        await expectLater(
          plugin.fetchCurrentDevice,
          throwsA(isA<NetworkException>()),
        );
      });
    });

    tearDown(() async {
      await plugin.close();
    });
  });
}
```

Comment (hahnandrew) [reply packages/auth/amplify_auth_cognito_test/test/plugin/fetch_current_device.dart]: [Added](https://github.com/aws-amplify/amplify-flutter/pull/5074/commits/1cbc9fa2fa68f74bbf98d7504efa52158ebd5118). I realize the issue I had was the previous `tearDown` from the first group, `setUp()` does properly scope.

Comment (hahnandrew) [reply packages/auth/amplify_auth_cognito_test/test/plugin/fetch_current_device.dart]: [Added](https://github.com/aws-amplify/amplify-flutter/pull/5074/commits/1cbc9fa2fa68f74bbf98d7504efa52158ebd5118)

File: packages/auth/amplify_auth_cognito_test/test/plugin/fetch_current_device_test.dart [hunk 4]
```diff
@@ -0,0 +1,101 @@
...
+      setUp(() async {
+        final mockIdp = MockCognitoIdentityProviderClient(
+          getDevice: () async => mockDeviceResponse,
+          forgetDevice: () async {},
+        );
+        plugin.stateMachine.addInstance<CognitoIdentityProviderClient>(mockIdp);
+      });
+
+      test(
+          'should get the current device. current device id should be equal to the local device id',
```
Comment (Equartey) [inline packages/auth/amplify_auth_cognito_test/test/plugin/fetch_current_device_test.dart:60]: nit: wording on the groups/tests do not flow. Currently, the test console would display: 
`fetchCurrentDevice Cognito GetDevice returns successfully should get the current device. current device id should be equal to the local device id`. Consider adjusting the wording to be consistent and read like a proper sentence.

Comment (hahnandrew) [reply packages/auth/amplify_auth_cognito_test/test/plugin/fetch_current_device_test.dart]: Added

File: packages/auth/amplify_auth_cognito_test/test/plugin/fetch_current_device_test.dart [hunk 5]
```diff
@@ -0,0 +1,99 @@
...
+        secureStorageFactory: (_) => secureStorage,
+      );
+      await plugin.configure(
+        config: mockConfig,
+        authProviderRepo: testAuthRepo,
+      );
+      repo = plugin.stateMachine.getOrCreate<DeviceMetadataRepository>();
+    });
+
+    group('Cognito GetDevice returns successfully.', () {
```
Comment (Equartey) [inline packages/auth/amplify_auth_cognito_test/test/plugin/fetch_current_device_test.dart:45]:
```suggestion
    group('should successfully', () {
```

File: packages/auth/amplify_auth_cognito_test/test/plugin/fetch_current_device_test.dart [hunk 6]
```diff
@@ -0,0 +1,99 @@
...
+      setUp(() async {
+        final mockIdp = MockCognitoIdentityProviderClient(
+          getDevice: () async => mockDeviceResponse,
+          forgetDevice: () async {},
+        );
+        plugin.stateMachine.addInstance<CognitoIdentityProviderClient>(mockIdp);
+      });
+
+      test(
+          'This test should get the current device, where the current device id is equal to the local device id',
```
Comment (Equartey) [inline packages/auth/amplify_auth_cognito_test/test/plugin/fetch_current_device_test.dart:55]:
```suggestion
          'return the current device where the current device id is equal to the local device id',
```

File: packages/auth/amplify_auth_cognito_test/test/plugin/fetch_current_device_test.dart [hunk 7]
```diff
@@ -0,0 +1,99 @@
...
+          () async {
+        await plugin.forgetDevice();
+        await expectLater(
+          plugin.fetchCurrentDevice,
+          throwsA(isA<DeviceNotTrackedException>()),
+        );
+      });
+    });
+
+    group('This test should have Cognito GetDevice throw a AWSHttpException',
```
Comment (Equartey) [inline packages/auth/amplify_auth_cognito_test/test/plugin/fetch_current_device_test.dart:75]:
```suggestion
    group('should throw',
```

File: packages/auth/amplify_auth_cognito_test/test/plugin/fetch_current_device_test.dart [hunk 8]
```diff
@@ -0,0 +1,99 @@
...
+          () async {
+        final secrets = await repo.get(username);
+        final currentDeviceKey = secrets?.deviceKey;
+        expect(currentDeviceKey, isNotNull);
+        final currentDevice = await plugin.fetchCurrentDevice();
+        expect(currentDeviceKey, currentDevice.id);
+      });
+
+      test(
+          'This test should throw a DeviceNotTrackedException when current device key is null',
```
Comment (Equartey) [inline packages/auth/amplify_auth_cognito_test/test/plugin/fetch_current_device_test.dart:65]: I think this test is also under the wrong group. Move it to the `should throw` group. 

```suggestion
          'a DeviceNotTrackedException when current device key is null',
```

**Review by Equartey — APPROVED** [review]

**Review by Jordan-Nelson — APPROVED** [review]
